libraries/designsystem/src/main/kotlin/io/element/android/libraries/designsystem/components/Bloom.kt

```diff
@@ -0,0 +1,534 @@
+}
+
+/**
+ * Bloom effect modifier for avatars. Applies a bloom effect to the component.
+ * @param avatarData The avatar data to use as the bloom source.
+ *  If the avatar data has a URL it will be used as the bloom source, otherwise the initials will be used. If `null` is passed, no bloom effect will be applied.
+ * @param background The background color to use for the bloom effect. Since we use blend modes it must be non-transparent.
+ * @param blurSize The size of the bloom effect. If not specified the bloom effect will be the size of the component.
+ * @param offset The offset to use for the bloom effect. If not specified the bloom effect will be centered on the component.
+ * @param clipToSize The size to use for clipping the bloom effect. If not specified the bloom effect will not be clipped.
+ * @param bottomSoftEdgeHeight The height of the bottom soft edge. If not specified the bottom soft edge will not be drawn.
+ * @param bottomSoftEdgeAlpha The alpha value to apply to the bottom soft edge.
+ * @param alpha The alpha value to apply to the bloom effect.
+ */
+fun Modifier.avatarBloom(
+    avatarData: AvatarData?,
+    background: Color,
+    blurSize: DpSize = DpSize.Unspecified,
+    offset: DpOffset = DpOffset.Unspecified,
+    clipToSize: DpSize = DpSize.Unspecified,
+    bottomSoftEdgeHeight: Dp = 40.dp,
+    @FloatRange(from = 0.0, to = 1.0)
+    bottomSoftEdgeAlpha: Float = 1.0f,
+    @FloatRange(from = 0.0, to = 1.0)
+    alpha: Float = 1f,
+) = composed {
+    // Bloom only works on API 29+
+    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) return@composed this
+    avatarData ?: return@composed this
+
+    if (avatarData.url != null) {
+        // Request the avatar contents to use as the bloom source
+        val painter = rememberAsyncImagePainter(
+            ImageRequest.Builder(LocalContext.current)
+                .data(avatarData)
+                // Needed to be able to read pixels from the Bitmap for the hash
+                .allowHardware(false)
+                .build()
+        )
+        var blurHash by remember { mutableStateOf<String?>(null) }
+
+        LaunchedEffect(avatarData) {
+            val drawable = painter.imageLoader.execute(painter.request).drawable ?: return@LaunchedEffect
+            val bitmap = (drawable as? BitmapDrawable)?.bitmap ?: return@LaunchedEffect
+            blurHash = BlurHash.encode(bitmap, BloomDefaults.HASH_COMPONENTS, BloomDefaults.HASH_COMPONENTS)
+        }
+
+        bloom(
+            hash = blurHash,
+            background = background,
+            blurSize = blurSize,
+            offset = offset,
+            clipToSize = clipToSize,
+            bottomSoftEdgeHeight = bottomSoftEdgeHeight,
+            bottomSoftEdgeAlpha = bottomSoftEdgeAlpha,
+            alpha = alpha,
+        )
+    } else {
+        // There is no URL so we'll generate an avatar with the initials and use that as the bloom source
+        val avatarColors = avatarColors(avatarData.id)
```

Comment:
I think this will not compile once merged on develop (just viewing this)

Reply:
Yes, I need to update this now the avatar's PR is merged. Thanks for the heads up though!

Reply:
Done in https://github.com/vector-im/element-x-android/pull/1253/commits/9d1bac54bc9b712d4dcd9c12b3e47eccf233ad3c.